Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,478 @@
+/-
+Copyright (c) 2023 Adomas Baliuka. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adomas Baliuka
+-/
+import Mathlib.Analysis.SpecialFunctions.Log.NegMulLog
+import Mathlib.Analysis.Convex.SpecificFunctions.Basic
+
+/-!
+# Properties of (Shannon) q-ary entropy and binary entropy functions
+
+The [binary entropy function](https://en.wikipedia.org/wiki/Binary_entropy_function)
+`binaryEntropy p := - p * log p - (1 - p) * log (1 - p)`
+is the Shannon entropy of a Bernoulli random variable with success probability `p`.
+
+More generally, the q-ary entropy function is the Shannon entropy of the random variable
+with possible outcomes `{1, ..., q}`, where outcome `1` has probability `(1 - p)`
+and all other outcomes are equally likely.
+
+`qaryEntropy (q : ℕ) (p : ℝ) := p * log (q - 1) - p * log p - (1 - p) * log (1 - p)`
+
+This file assumes that entropy is measured in Nats, hence the use of natural logarithms.
+Most lemmas are also valid using a different-base logarithms.
+
+## Tags
+
+entropy, Shannon, binary
+-/
+
+namespace Entropy
+
```

Comment:
I heard there was a trend towards namespacing everything in Mathlib (which I would naively be in favor of). Is there some higher-level namespace this could/should go into? Something like `SpecialFunctions`?

Reply:
There was some talk of namespacing *all* of mathlib for the sake of interoperability with other libraries, but that's a different story. The policy hasn't really changed: if there were a risk that `binaryEntropy` and `qaryEntropy` would have different meanings in different areas of math, then namespacing would definitely make sense. But I can't really see this happening here.